f3dex2.s

```diff
@@ -222,16 +250,57 @@ geometryModeLabel:
 // excluding ambient light
 MAX_LIGHTS equ 7
 
-// 0x01F0-0x021F: Light data
-lightBuffer:
-    .fill ((MAX_LIGHTS + 1) * 6)
-
-// 0x0220-0x0240: Light colors
-lightColors:
-    .fill ((MAX_LIGHTS + 1) * 4)
-
-// 0x0240-0x02E0: ??
-.skip 0xA0
+// 0x01F0-0x02E0: Light data; a total of 10 * lightSize light slots.
+// Each slot's data is either directional or point (each pair of letters is a byte):
+//      Directional lights:
+// 0x00 RR GG BB 00 RR GG BB -- NX NY NZ -- -- -- -- --
+// 0x10 TX TY TZ -- TX TY TZ -- (Normals transformed to camera space)
+//      Point lights: 
+// 0x00 RR GG BB CC RR GG BB LL XXXX YYYY ZZZZ QQ --
+// 0x10 -- -- -- -- -- -- -- -- (Invalid transformed normals get stored here)
+// CC: constant attenuation factor (0 indicates directional light)
+// LL: linear attenuation factor
+// QQ: quadratic attenuation factor
+//
+// First there are two lights, whose directions define the X and Y directions
+// for texgen, via g(s)SPLookAtX/Y. The colors are ignored. These lights get
+// transformed normals. g(s)SPLight which point here start copying at n*24+24,
+// where n starts from 1 for one light (or zero lights), which effectively
+// points at lightBufferMain.
+lightBufferLookat:
+    .fill (2 * lightSize)
+// Then there are the main 8 lights. This is between one and seven directional /
+// point (if built with this enabled) lights, plus the ambient light at the end.
+// Zero lights is not supported, and is encoded as one light with black color
+// (does not affect the result). Once there is one point light, the rest (until
+// ambient) are also assumed to be point lights.
+lightBufferMain:
+    .fill (8 * lightSize)
+// The addressing of all of this is convoluted and took some thought to work
```

Comment:
While I do find the explanation of how this was worked out very interesting, I don't think we need to explain this in the code itself. My concern is that this file could get very long if every reverse engineering effort for this microcode had an explanation like this, so I'd like to keep explanations of the process like this one out of the code.

Reply:
Got it, removed.